basilica/basilica/__init__.py

```diff
@@ -81,16 +86,60 @@ def raw_embed(self, url, data, opts, timeout):
     # TODO: parallelize
     def embed(self, url, data, batch_size, opts, timeout):
         batch = []
+        outstanding = 0
+        api_thread = threading.Thread(target=self.raw_embed_wrapper, args=(url, opts, timeout))
+        api_thread.setDaemon(True)
+        api_thread.start()
         for i in data:
             batch.append(i)
             if len(batch) >= batch_size:
-                for e in self.raw_embed(url, batch, opts=opts, timeout=timeout):
-                    yield e
+                self.write_queue.put_nowait(batch)
                 batch = []
+                if self.thread_running:
+                    emb = self.read_queue.get(block=True)
+                    if isinstance(emb, Exception):
+                        raise emb
+                    for e in emb:
+                        yield e
+                try:
+                    emb = self.read_queue.get(block=False) 
+                except Queue.Empty:
+                    continue
+                else:
+                    for e in emb:
+                        yield e
         if len(batch) > 0:
-            for e in self.raw_embed(url, batch, opts=opts, timeout=timeout):
-                yield e
-            batch = []
+            self.write_queue.put_nowait(batch)
+        while not self.write_queue.empty() or self.thread_running:
+            try:
+                emb = self.read_queue.get(block=True)
+                if isinstance(emb, Exception):
+                    raise emb
+            except Queue.Empty:
+                continue
+            except Exception as e:
+                raise e
+            else:
+                for e in emb:
+                    yield e
+
+    def raw_embed_wrapper(self, url, opts, timeout):
+        while True:
+            try:
+                batch = self.write_queue.get(block=False)
+                self.thread_running = True
```

Comment:
I don't think we want to be using a variable for coordination across threads like this.  If we do want to, we should lock before accessing it, but I think we probably want to be using the queue for cross-thread coordination.

Reply:
I read through the queue documentation.  I think a reasonable way to get what we want is to set the size of the write queue to 1, and replace the `put_nowait`s with regular `put`s, and stop tracking `thread_running` at all.  That way the `put`s will block until the worker thread has caught up, which will provide the backpressure we want.

Reply:
If we do this we'll also have to pass an "I'm done" message over the read queue (see comment below about passing one for the write queue) so that we know when to terminate.